### PR TITLE
Refactored `DeviceCache` to improve thread safety

### LINKED ISSUE #739: DeviceCache thread safety
DeviceCache has mechanisms for thread safety, and many objects that have a reference to the DeviceCache also control access internally, however, if you have multiple objects each on different threads reading and then writing data, there might be times when a race condition can develop. DeviceCache uses a queue to add safety. However, there is nothing stopping a situation like:
Thread 1 reads data from cache
Thread 2 reads data from cache
Thread 1 writes data back to cache
Thread 2 writes stale data back to cache overwriting changes

Most places where we manipulate the cache, we rely on the objects utilizing the cache to control access by:

1.  locking
2. reading
3. writing
4. unlocking

However, because no objects share locks, we can still get into this race condition situation because the Device Cache is ready to service read/write requests from a different thread as soon as `2.` happens. 

**How to fix?**

Maybe something like how `CoreData` manages this type of situation? The `ManagedObjectContext` has a `perform` function that executes a block of code in the context's queue. If we were to do something like that, we could ensure steps `1` through `4` all happened atomically.
This would change the API so that device cache looked more like:

```swift
deviceCache.perform { cache in
    let someValues = cache.read()
    ...
    cache.write(modifiedValues)
}
```

### PR DESCRIPTION
**Pull request by NachoSoto**
Fixes #739.

### Changes:
- Created `Atomic`, based on `Lock`, to abstract atomic access and modifications to values.
- Simplified `InMemoryCachedObject` using `Atomic` (tests still pass!).
- Created `SynchronizedUserDefaults` using `Atomic`, to enforce atomic access to `UserDefaults`.
- Refactored `DeviceCache` using `Atomic` to enforce all access to `UserDefaults` are synchronized:
  - `DeviceCache` no longer stores a `UserDefaults` instance (only `SynchronizedUserDefaults`), so we guarantee we don't do unsynchronized modifications.
  - The old `threadUnsafe_` methods are now `static`, which ensures they don't access any local state. Instead, `UserDefaults` is injected, and `DeviceCache` calls these methods using `SynchronizedUserDefaults.perform`
  - The combination of both of these things eliminates potential for programmer error within the class.
  - There's still potential for what's described in #739, that a _user_ of `DeviceCache` reads data, and THEN writes it again, without synchronizing both operations. This PR doesn't address that, but I haven't found any examples of that being an issue, and if we do, we should move both operations into a single method in `DeviceCache` that can enforce the semantics.

### CODE REVIEW

**Review by taquitos — COMMENTED** [review]
I think this is great. Before we merge, I'd love to see a new markdown doc in `docs/` that describes the relationships between `Atomic`, `Lock`, and `UserDefaults`. I'm thinking something like an ELI5- keeping in mind that as we onboard junior engineers, they likely won't have a lot of exposure to generics/atomicity.

File: Purchases/Misc/Atomic.swift [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+//      https://opensource.org/licenses/MIT
+//
+//  Atomic.swift
+//
+//  Created by Nacho Soto on 11/20/21.
+
+import Foundation
+
+/// An atomic variable.
+internal final class Atomic<T> {
```
Comment (aboedo) [inline Purchases/Misc/Atomic.swift:17]: nitpick: I kinda feel like if we improve the class description here, the comments on each operation could be removed. 
like, if we had something like: 
```swift
/** A generic object that performs all write and read operations atomically. 
* Use it to prevent data races when accessing an object. 
* 
* Usage: 
* ```swift
* let foo = Atomic<MyClass>
* // read values
* foo.withValue { 
*     currentBar = $0.bar
*     currentX = $0.x
* }
* // write value
* foo.modify { 
*     $0.bar = 2
*     $0.x = "new X"
* }
* ```
* 
* Or for single-line read/writes: 
* ```swift
* let currentX = foo.value.x
* foo.value = MyClass()
* ``` 
**/
```

Then the usage and even parts of the implementation might seem clearer on first glance, and the other comments would be redundant, since we know all operations are performed atomically, from the class name and the docs

**Review by aboedo — COMMENTED** [review]
This is absolutely brilliant. 
Perhaps we should add tests for `Atomic` to ensure we don't break it in the future? I get the feeling that a future dev could easily mess up if they ever decide to modify it

File: Purchases/Misc/Atomic.swift (hunk 1 above)
Comment (NachoSoto) [reply Purchases/Misc/Atomic.swift:17]: Oh that's a great idea! Much easier to understand than poorly worded explanations of each method 😛

Comment (NachoSoto) [reply Purchases/Misc/Atomic.swift:17]: Oh that's a great idea! Much easier to understand than poorly worded explanations of each method 😛

**Comment by NachoSoto** [comment]
Thanks for the review!

> Perhaps we should add tests for Atomic to ensure we don't break it in the future? 
Totally :+1: will do

> Before we merge, I'd love to see a new markdown doc in docs/ that describes the relationships between Atomic, Lock, and UserDefaults
Great idea! Will do.

> I'm thinking something like an ELI5
`ELI5`?

**Comment by taquitos** [comment]
ELI5: Literally: Explain like I'm five years old: https://www.reddit.com/r/explainlikeimfive/
Making the explanation as easy as possible for folks who have pretty much no context 😄

**Comment by aboedo** [comment]
that explains a lot 😂 I googled ELI5 and [this python library](https://eli5.readthedocs.io/en/latest/) popped up, couldn't figure out how that would apply here

**Comment by NachoSoto** [comment]
> I think this is great. Before we merge, I'd love to see a new markdown doc in `docs/` that describes the relationships between `Atomic`, `Lock`, and `UserDefaults`. I'm thinking something like an ELI5- keeping in mind that as we onboard junior engineers, they likely won't have a lot of exposure to generics/atomicity.

Done! Let me know what you think.

**Comment by NachoSoto** [comment]
> This is absolutely brilliant! ⭐ ✨
> Perhaps we should add tests for `Atomic` to ensure we don't break it in the future? I get the feeling that a future dev could easily mess up if they ever decide to modify it

Done. I've only included very simple tests that don't cover thread-safety itself. Such tests can be brittle and, if the implementation or the test is incorrect, lead to sporadic failures. But let me know if you think we should still have something to cover that.

File: PurchasesTests/Misc/AtomicTests.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+//
+//  Created by Nacho Soto on 11/25/21.
+
+import Nimble
+import XCTest
+
+@testable import RevenueCat
+
+class AtomicTests: XCTestCase {
+    func testInitialValue() {
```
Comment (aboedo) [inline PurchasesTests/Misc/AtomicTests.swift:20]: suggestion: `testInitSetsValueCorrectly`

File: PurchasesTests/Misc/AtomicTests.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        let value = Int.random(in: 0..<100)
+        let atomic = Atomic(value)
+
+        expect(atomic.value) == value
+    }
+
+    func testModify() {
```
Comment (aboedo) [inline PurchasesTests/Misc/AtomicTests.swift:27]: suggestion: `testModifyUpdatesTheValue`

File: PurchasesTests/Misc/AtomicTests.swift [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    func testModifyReturnsResult() {
+        let atomic = Atomic(10)
+        let result: Bool = atomic.modify { $0 += 10; return false }
+
+        expect(result) == false
+    }
+
+    func testWithValue() {
```
Comment (aboedo) [inline PurchasesTests/Misc/AtomicTests.swift:41]: suggestion: `testWithValueUsesTheCorrectValueWithinTheBlock`

File: docs/ThreadSafety.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
+# Thread Safety 👾
```
Comment (aboedo) [inline docs/ThreadSafety.md:1]: ❤️

**Review by aboedo — APPROVED** [review]
looks awesome! Thanks for the updates!

**Comment by NachoSoto** [comment]
Thanks for the suggestions! I like adding the expected results in the test name 👍
